### PR TITLE
feat: (Core) introduce Rating Indicator component 

### PR DESCRIPTION
**Pull request by mykolavorobiov**
#### Please provide a link to the associated issue.
https://github.com/SAP/fundamental-ngx/issues/3037

#### Please provide a brief summary of this pull request.
Introduce Rating Indicator component to `fundumental-ngx` core
Note: original styles with small improvement implement as part of a component. We need to fix this issue: https://github.com/SAP/fundamental-styles/issues/1753
After that, we can apply normal import from style package, like this `@import '~fundamental-styles/dist/rating-indicator';`

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples

### CODE REVIEW

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-base.component.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { Component } from '@angular/core';
+
+@Component({
+    selector: 'app-rating-indicator-base',
+    templateUrl: './rating-indicator-base.component.html',
+    styleUrls: ['./rating-indicator-style.component.scss']
+})
+export class RatingIndicatorBaseComponent {
+    onRatingChanged(event: number): void {
+        console.log(event)
```
Comment (katekozlowska) [inline apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-base.component.ts:10]: remove log

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-examples.component.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        indicatorTotal: 5,
+        allowHalves: false,
+        disabled: false,
+        size: 'md'
+    };
+    sizes = ['xs', 'sm', 'md', 'lg', 'cozy', 'compact', 'condensed'];
+
+    isDisable = false;
+
+    onRatingChanged(event: number): void {
```
Comment (katekozlowska) [inline apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-examples.component.ts:19]: do you need this function?

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-style.component.scss [hunk 3]
```diff
@@ -0,0 +1,12 @@
+.base-row {
+    display: flex;
+    align-items: flex-start;
+    justify-content: flex-start;
+    flex-direction: column;
+    padding-bottom: 0.5rem;
+    margin-bottom: 0.5rem;
+
+    &:not(:last-child) {
+        border-bottom: 1px solid grey;;
```
Comment (katekozlowska) [inline apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-style.component.scss:10]: remove double mark :)

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-examples.component.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        console.log(event);
+    }
+
+    onRatingChangedWithDisable(event: number): void {
+        console.log(event);
```
Comment (katekozlowska) [inline apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-examples.component.ts:24]: please remove it

**Review by katekozlowska — CHANGES_REQUESTED** [review]

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-base.component.ts (hunk 1 above)
Comment (mykolavorobiov) [reply apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-base.component.ts:10]: It's an example. `log` is not really a helpful example, maybe we have a default flow to show event data. If yes, share it with me, please.

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-examples.component.ts (hunk 4 above)
Comment (mykolavorobiov) [reply apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-examples.component.ts:24]: Same as above

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-examples.component.ts (hunk 2 above)
Comment (mykolavorobiov) [reply apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-examples.component.ts:19]: It's an example only.

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    @HostBinding('class.is-display-mode')
+    displayMode = false;
+
+    /**
+     * Number of stars/icons/images to display. Max value is 7 and min valud is 3
+     */
+    @Input()
+    set indicatorTotal(value: number) {
+        if (isNaN(+value)) {
+            this._indicatorTotal = 5;
```
Comment (JKMarkowski) [inline libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:72]: Maybe let's move `5` to constants

**Review by JKMarkowski — CHANGES_REQUESTED** [review]
Great Job! I added some comments regarding code. Consider using `ControlValueAccessor` there. Also you need to listen no changes in inputs triggered by arrows. Othwerise model is not the same as view 
rating indicator

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-style.component.scss (hunk 3 above)
Comment (mykolavorobiov) [reply apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-style.component.scss]: Removed

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts (hunk 5 above)
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:72]: Yep, will add it

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.html [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        </label>
+    </ng-container>
+</div>
+<!-- Popover if displaySumPopover is true and we have ratings object -->
+<fd-popover
+    [triggers]="['mouseenter', 'mouseleave']"
+    [noArrow]="false"
+    placement="top"
+    addContainerClass="fd-rating-indicator__popover"
+    *ngIf="displaySumPopover && _ratingItems && _ratingItems.length; else dynamicText"
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.html:33]: Yeah, will change

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    @Input()
+    @HostBinding('class.is-display-mode')
+    displayMode = false;
+
+    /**
+     * Number of stars/icons/images to display. Max value is 7 and min valud is 3
+     */
+    @Input()
+    set indicatorTotal(value: number) {
+        if (isNaN(+value)) {
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:71]: Not really needed, will remove

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+            this._hideDynamicText || !this._value ? INDICATOR_CLASSES.hideDynamicText : '',
+            this.class
+        ];
+    }
+
+    /** @hidden */
+    private _generateRatings(): void {
+        if (this.ratings) {
+            let sum = 0;
+            let allVoites = 0;
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:243]: Thx, will change

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+            this._ratingItems = items;
+            this.ratingAverage = sum / allVoites * 10;
+            this.totalRatings = allVoites;
+            this._value = this._convertToValue();
+        }
+
+    }
+
+    /** @hidden */
+    private _convertToValue(): number {
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:271]: Okay, will add

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+        return this.value ? this._parseValue(+this.value) : this._parseValue(this.ratingAverage);
+    }
+
+    /** @hidden */
+    private _parseValue(value: number): number {
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:276]: Okay, will add

File: libs/core/src/lib/rating-indicator/constants.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
+export const INDICATOR_PREFIX = 'fd-rating-indicator';
+
+export const INDICATOR_CLASSES = {
+    halves: `${INDICATOR_PREFIX}--half-star`,
+    icon: `${INDICATOR_PREFIX}--icon`,
+    hideDynamicText: `${INDICATOR_PREFIX}--hide-dynamic-text`
+};
+export const INDICATOR_DEFAULT_TOTAL = 5;
+export const INDICATOR_RANGE = {
```
Comment (salarenko) [inline libs/core/src/lib/rating-indicator/constants.ts:9]: This range is specified in the docs?

**Review by salarenko — CHANGES_REQUESTED** [review]
- [x] I don't think we need to double example title
image

- [x] Example components should use following naming convention: `[COMPONENT_NAME]ExampleComponent`

- [ ] Check how many times are you casting the `value` with `+value`. Won't it be better to just make sure `value` is a number once when the value is set, and later use it safely?

- [ ] There is a state inconsistency. Indicator control cannot visually represent value different than the value stored in the control. Where does this "press space or enter" functionality comes from?
rating-inconsistenct

- [ ] Navigating with `ArrowUp` and `ArrowDown` does not open the popover with the message.
rating-up-down

- [ ] Try escaping rating indicator with arrows to the left or right and check how the popover with the message behaves.
rating-edges

- [ ] Missng tests of `RatingIndicatorComponent` class methods

File: libs/core/src/lib/rating-indicator/constants.ts (hunk 11 above)
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/constants.ts:9]: Nope, same as not specified ngModel and any control details for this component

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+    /** @hidden */
+    ngOnInit(): void {
+        this._value = this._convertToValue();
+        this._icons = this._getRates();
+        this.buildComponentCssClass();
+        this._generateRatings();
+        this._navigationArrowNoteSub = this.$navigationArrowNote.pipe(
+            debounceTime(100),
+            distinctUntilChanged(),
+            tap({
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:209]: Yep, anyway I'll remove this code and popover functionality for arrow buttons. It's not a good idea without specification.

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 13]
```diff
@@ -0,0 +1,381 @@
...
+    navigationNote = 'Please press enter/space for the vote.';
+
+    /**
+     * Fired when the user sets or changes their rating.
+     */
+    @Output()
+    ratingChanged = new EventEmitter<number>();
+
+    /** @hidden */
+    @ViewChild(PopoverComponent, { static: false })
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:151]: Yep, anyway I'll remove popover functionality for arrow buttons. It's not a good idea without specification.

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 14]
```diff
@@ -0,0 +1,381 @@
...
+     */
+    @Input()
+    @HostBinding('class.is-display-mode')
+    displayMode = false;
+
+    /**
+     * Number of stars/icons/images to display. Max value is 7 and min valud is 3
+     */
+    @Input()
+    set indicatorTotal(value: number) {
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:78]: Yep, got it.
p.s. copy from tech doc
image

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+        if (isNaN(value)) {
+            this._indicatorTotal = INDICATOR_DEFAULT_TOTAL;
+        } else {
+            this._indicatorTotal = Math.max(INDICATOR_RANGE.min, Math.min(INDICATOR_RANGE.max, +value));
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:82]: I don't want to add an additional function from cdk, but agree with you, it's really better than my version.
Will update it.
p.s. I don't trust the user.

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 16]
```diff
@@ -0,0 +1,381 @@
...
+    /** @hidden */
+    ngOnChanges(changes: SimpleChanges): void {
+        if ('class' in changes || 'size' in changes) {
+            this.buildComponentCssClass();
+        }
+        if ('value' in changes) {
+            this._value = this._convertToValue();
+        }
+        if ('indicatorTotal' in changes || 'allowHalves' in changes) {
+            this._value = this._convertToValue();
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:230]: Yep, it's still needed, if you change the `allowHalves` option.

https://monosnap.com/file/3S8qiUKWJTjgrF994GsaTFnKJIbtx5

[the remark this answers is not in the document]
Comment (JKMarkowski) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:224]: Do you need to keep `indicatorCapacity` there as well?

**Review by JKMarkowski — CHANGES_REQUESTED** [review]
Hi @mykolavorobiov I added some comments regarding examples/code, after addressing we'll be good to merge

File: libs/core/tsconfig.lib.json [hunk 17]
```diff
@@ -12,7 +12,7 @@
         "experimentalDecorators": true,
         "importHelpers": true,
         "types": [],
-        "lib": ["dom", "es2015"]
+        "lib": ["dom", "es2018"]
```
Comment (JKMarkowski) [inline libs/core/tsconfig.lib.json:15]: Is there any reason for this change?

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-example.component.html [hunk 18]
```diff
@@ -0,0 +1,119 @@
+<div class="base-row">
+    <fd-rating-indicator
+        indicatorCapacity="5"
+        size="xsssss"
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-example.component.html:4]: I protected this field with predefined sizes, and it's an example only.

File: apps/docs/src/app/core/component-docs/rating-indicator/rating-indicator-docs.component.ts [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    selector: 'app-rating-indicator',
+    templateUrl: './rating-indicator-docs.component.html'
+})
+export class RatingIndicatorDocsComponent {
+    ratingIndicatorSizes: ExampleFile[] = [
+        {
+            language: 'html',
+            code: sizeRatingHtml,
+            fileName: 'rating-indicator-sizes',
+            typescriptFileCode: sizesRatingTs,
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply apps/docs/src/app/core/component-docs/rating-indicator/rating-indicator-docs.component.ts:23]: Thx, will fix

File: apps/docs/src/app/core/component-docs/rating-indicator/rating-indicator-docs.component.ts [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+            component: 'RatingIndicatorSizesExampleComponent'
+        }
+    ];
+
+    ratingIndicatorRatings: ExampleFile[] = [
+        {
+            language: 'html',
+            code: ratingsObjectHtml,
+            fileName: 'rating-indicator-ratings',
+            typescriptFileCode: ratingsObjectTs,
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply apps/docs/src/app/core/component-docs/rating-indicator/rating-indicator-docs.component.ts:33]: Thx, will fix

File: apps/docs/src/app/core/component-docs/rating-indicator/rating-indicator-docs.component.ts [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    ];
+
+    ratingIndicatorFewTest: ExampleFile[] = [
+        {
+            language: 'html',
+            code: baseHtml,
+            fileName: 'rating-indicator-example',
+            typescriptFileCode: baseTs,
+            component: 'RatingIndicatorExampleComponent'
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply apps/docs/src/app/core/component-docs/rating-indicator/rating-indicator-docs.component.ts:44]: Thx, will fix

File: libs/core/tsconfig.lib.json (hunk 17 above)
Comment (mykolavorobiov) [reply libs/core/tsconfig.lib.json:15]: I want to use `Object.entries`, and don't want to use it without a polyfill. Do we have blockers to update the version?

Also, I saw an issue from @frankfangcn https://github.com/SAP/fundamental-ngx/issues/3710

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.html [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+        </span>
+    </fd-popover-control>
+    <fd-popover-body>
+        <div class="rate-sum-row" *ngFor="let rating of _ratingItems; trackBy: trackByFn">
+            <div class="fd-rating-indicator__container rates">
+                <ng-container *ngFor="let rate of _rates; trackBy: trackByFn">
+                    <span [class.fd-rating-indicator__input-checked]="rate.value === rating.rate" class="fd-rating-indicator__input"></span>
+                    <label class="fd-rating-indicator__label">
+                        <i *ngIf="ratedIcon" class="rated" [ngClass]="unratedIcon"></i>
+                        <i *ngIf="unratedIcon" class="unrated" [ngClass]="ratedIcon"></i>
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.html:41]: Nope, it's a hack for correct working with `~` in styles

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-example.component.html [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+    (ratingChanged)="onRatingChanged($event)"
+></fd-rating-indicator>
+<h2>Custom class for icon</h2>
+<fd-rating-indicator
+    totalRatings="1"
+    value="3.34"
+    [allowHalves]="true"
+    ratedIcon="sap-icon--bo-strategy-management"
+    unratedIcon="sap-icon--notification"
+></fd-rating-indicator>
```
[the remark this answers is not in the document]
Comment (mykolavorobiov) [reply apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-example.component.html:86]: Yep, will do it

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-ratings.component.html [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+    <fd-rating-indicator
+        indicatorCapacity="5"
+        size="xs"
+        [displaySumPopover]="true"
+        [ratings]="testRatingData1"
+        #rate1
+    ></fd-rating-indicator>
+    <p>
+        Rate average: {{ rate1.ratingAverage.toFixed(2) }}
+        Total ragings: {{ rate1.totalRatings }}
```
Comment (stefanoScalzo) [inline apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-ratings.component.html:13]: typo

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-ratings.component.html [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+    <fd-rating-indicator
+        indicatorCapacity="7"
+        size="compact"
+        [displaySumPopover]="true"
+        [ratings]="testRatingData2"
+        #rate2
+    ></fd-rating-indicator>
+    <p>
+        Rate average: {{ rate2.ratingAverage.toFixed(2) }}
+        Total ragings: {{ rate2.totalRatings }}
```
Comment (stefanoScalzo) [inline apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-ratings.component.html:29]: typo

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-ratings.component.html [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+    <fd-rating-indicator
+        indicatorCapacity="5"
+        size="lg"
+        [displaySumPopover]="true"
+        [ratings]="testRatingData3"
+        #rate3
+    ></fd-rating-indicator>
+    <p>
+        Rate average: {{ rate3.ratingAverage.toFixed(2) }}
+        Total ragings: {{ rate3.totalRatings }}
```
Comment (stefanoScalzo) [inline apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-ratings.component.html:45]: typo

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-ratings.component.html [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+    <fd-rating-indicator
+        indicatorCapacity="5"
+        size="lg"
+        [displaySumPopover]="true"
+        [ratings]="testRatingData4"
+        #rate4
+    ></fd-rating-indicator>
+    <span>
+        Rate average: {{ rate4.ratingAverage.toFixed(2) }}
+        Total ragings: {{ rate4.totalRatings }}
```
Comment (stefanoScalzo) [inline apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-ratings.component.html:61]: typo

**Comment by stefanoScalzo** [comment]
we need to add a lot more documentation on when to use each use case for each example and how our API works with it.
For the Rating indicator examples, rating with the number input, why is the number increasing after half, like it might make sense but the user can't see the code.

**Comment by mykolavorobiov** [comment]
> we need to add a lot more documentation on when to use each use case for each example and how our API works with it.
> For the Rating indicator examples, rating with the number input, why is the number increasing after half, like it might make sense but the user can't see the code.

@stefanoScalzo 
Adding more examples is not a problem, but maybe you can help with this and share with me what're examples I could add?
About number input - I don't understand what do you mean? What's the code need to show to the user?
About number after half - I don't change the initial value, only view value. Unfortunately, in our styles implementation, we can use whole or with halves stars only.
And it means if the user wants to vote, he could use only a predefined value.
If you mean that I need to add some explanation about half and number after the half, please let me know.

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.html [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+            [checked]="viewValue === rate.value"
+            (focus)="onSelect(rate.value)" />
+        <label class="fd-rating-indicator__label" [for]="rate.id">
+            <i *ngIf="ratedIcon" class="rated" [ngClass]="unratedIcon"></i>
+            <i *ngIf="unratedIcon" class="unrated" [ngClass]="ratedIcon"></i>
+        </label>
+    </ng-container>
+</div>
+<!-- Popover if displaySumPopover is true and we have ratings object -->
+<fd-popover
```
Comment (salarenko) [inline libs/core/src/lib/rating-indicator/components/rating-indicator.component.html:23]: I think we should switch to [this popover markup](https://fundamental-ngx.netlify.app/#/core/popover#trigger) to make the code more readable.

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 29]
```diff
@@ -0,0 +1,373 @@
...
+            multi: true
+        }
+    ]
+})
+export class RatingIndicatorComponent implements OnInit, OnChanges, CssClassBuilder, ControlValueAccessor {
+    /** User's custom classes */
+    @Input()
+    class: string;
+
+    /** Sets [name] property of input. */
```
Comment (salarenko) [inline libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:67]: A very detail, but "attribute" not "property" 😋

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 30]
```diff
@@ -0,0 +1,373 @@
...
+    @Input()
+    name: string;
+
+    /** User's aria-label attr */
```
Comment (salarenko) [inline libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:71]: Lets switch to: "Sets the `aria-label` attribute to the element."

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 31]
```diff
@@ -0,0 +1,373 @@
...
+     * Overrides totalRatings and ratingAverage.
+     */
+    @Input()
+    ratings: NumberKey<number>;
+
+    /**
+     * Whether or not to display the popover that shows the sum of each rating. Requires [ratings] object.
+     */
+    @Input()
+    displaySumPopover = false;
```
Comment (salarenko) [inline libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:136]: Maybe lets name it `displayTotalRatings` or `displayAllRatings`

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 32]
```diff
@@ -0,0 +1,373 @@
...
+
+    /**
+     * Icon class from fundamental-styles lib https://sap.github.io/fundamental-styles/?path=/docs/components-icon--sizes
+     */
+    @Input()
+    ratedIcon: string;
+
+    @Input()
```
Comment (salarenko) [inline libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:144]: Missing documentation annotation.

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 33]
```diff
@@ -0,0 +1,373 @@
...
+    ratings: NumberKey<number>;
+
+    /**
+     * Whether or not to display the popover that shows the sum of each rating. Requires [ratings] object.
+     */
+    @Input()
+    displaySumPopover = false;
+
+    /**
+     * Icon class from fundamental-styles lib https://sap.github.io/fundamental-styles/?path=/docs/components-icon--sizes
```
Comment (salarenko) [inline libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:139]: I think we should simplify to: "Rating icon class". Developers can use custom icons outside of Fundamental Styles library.

File: libs/core/src/lib/rating-indicator/constants.ts [hunk 34]
```diff
@@ -0,0 +1,23 @@
+export const INDICATOR_PREFIX = 'fd-rating-indicator';
+
+export const INDICATOR_CLASSES = {
+    halves: `${INDICATOR_PREFIX}--half-star`,
+    icon: `${INDICATOR_PREFIX}--icon`,
+    hideDynamicText: `${INDICATOR_PREFIX}--hide-dynamic-text`
+};
+export const INDICATOR_DEFAULT_CAPACITY = 5;
+export const INDICATOR_CAPACITY_RANGE = {
```
Comment (salarenko) [inline libs/core/src/lib/rating-indicator/constants.ts:9]: This default rage comes from the specification?

**Review by salarenko — CHANGES_REQUESTED** [review]
Nice improvements :)

File: libs/core/src/lib/rating-indicator/constants.ts (hunk 11 above)
Comment (salarenko) [reply libs/core/src/lib/rating-indicator/constants.ts]: Then I'm not sure if we should specify some fixed boundaries for the capacity of the Rating Indicator. I think its good to have this, but `<3;7>` is quite narrow. I will bring this to one of the meetings to see what the rest of the team thinks.

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts (hunk 33 above)
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:139]: Yep, but tech doc and style for this component is not supported a simple way for a custom icon. We need to change the structure and in my opinion, it would be simpler with two templates for a rated and unrated icon (for example 
image
)

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.html (hunk 28 above)
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.html:23]: Trigger directive hasn't any options for custom popover (triggers, noArrow, additionalBodyClass etc.)

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts (hunk 31 above)
Comment (mykolavorobiov) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:136]: Copied from tech spec...changed to `displayAllRatings`

File: apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-example.component.html [hunk 35]
```diff
@@ -0,0 +1,73 @@
+<div class="base-row">
+    <fd-rating-indicator
+        indicatorCapacity="5"
+        size="xsssss"
+        value="0.1"
+        i18n-dynamicTextIndicator
```
Comment (mikerodonnell89) [inline apps/docs/src/app/core/component-docs/rating-indicator/examples/rating-indicator-example.component.html:6]: what is this line doing? Can't we just use dynamicTextIndicator input?

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/rating-indicator/constants.ts (hunk 11 above)
Comment (salarenko) [reply libs/core/src/lib/rating-indicator/constants.ts]: @mykolavorobiov We had a conversation on this topic and we decided not to specify the boundaries for the loading indicator capacity.

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.html (hunk 28 above)
Comment (salarenko) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.html:23]: @mykolavorobiov Yeah, you still put all of the popover configuration on `fd-popover` element.

Comment (salarenko) [reply libs/core/src/lib/rating-indicator/components/rating-indicator.component.html:23]: @mykolavorobiov Yeah, you still put all of the popover configuration on fd-popover element.

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.spec.ts [hunk 36]
```diff
@@ -0,0 +1,102 @@
+import { ElementRef } from '@angular/core';
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { RatingIndicatorComponent } from './rating-indicator.component';
+
+const prefix = 'fd-rating-indicator';
+
+fdescribe('RatingIndicatorComponent', () => {
```
Comment (salarenko) [inline libs/core/src/lib/rating-indicator/components/rating-indicator.component.spec.ts:8]: Excluded tests!

**Review by JKMarkowski — APPROVED** [review]
LGTM. Great Job!

**Review by salarenko — APPROVED** [review]

**Comment by stefanoScalzo** [comment]
> > we need to add a lot more documentation on when to use each use case for each example and how our API works with it.
> > For the Rating indicator examples, rating with the number input, why is the number increasing after half, like it might make sense but the user can't see the code.
> 
> @stefanoScalzo
> Adding more examples is not a problem, but maybe you can help with this and share with me what're examples I could add?
> About number input - I don't understand what do you mean? What's the code need to show to the user?
> About number after half - I don't change the initial value, only view value. Unfortunately, in our styles implementation, we can use whole or with halves stars only.
> And it means if the user wants to vote, he could use only a predefined value.
> If you mean that I need to add some explanation about half and number after the half, please let me know.

For each of the examples you need to say what the inputs correspond to example
        size="xsssss"
        value="0.1"
        dynamicTextIndicator="ooofff"

Say what size will change, what value will change, etc.

File: .vscode/settings.json [hunk 37]
```diff
@@ -1,5 +1,5 @@
 {
-    "editor.formatOnSave": true,
+    "editor.formatOnSave": false,
```
Comment (Betrozov) [inline .vscode/settings.json:2]: it should be fixed by another PR

File: libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts [hunk 38]
```diff
@@ -0,0 +1,380 @@
...
+    }
+
+    /** @hidden */
+    writeValue(value: number): void {
+        this._value = this._parseValue(value);
+        this._changeDetectorRef.markForCheck();
+    }
+
+    /** @hidden */
+    registerOnChange(fn: any): void {
```
Comment (Betrozov) [inline libs/core/src/lib/rating-indicator/components/rating-indicator.component.ts:249]: should it be `Function`?

File: .vscode/settings.json (hunk 37 above)
Comment (mykolavorobiov) [reply .vscode/settings.json:2]: Nope, it shouldn't be in repo in general

Comment (Betrozov) [reply .vscode/settings.json:2]: https://github.com/SAP/fundamental-ngx/pull/3768/files#r520124796

**Review by Betrozov — APPROVED** [review]

File: .vscode/settings.json (hunk 37 above)
Comment (mykolavorobiov) [reply .vscode/settings.json]: Removed

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
Let's just show the typescript in the example for the "Rating indicator with predefined ratings object" example, otherwise looks good

**Comment by mykolavorobiov** [comment]
> Let's just show the typescript in the example for the "Rating indicator with predefined ratings object" example, otherwise looks good

Could you please clarify, what do you mean? I show defined data in template.
image

**Review by mikerodonnell89 — APPROVED** [review]
